src/ZipkinBundle/Components/HttpClient/HttpClient.php

```diff
@@ -0,0 +1,139 @@
+
+        $headers = [];
+        ($this->injector)($span->getContext(), $headers);
+
+        try {
+            $options['headers'] = $headers + ($options['headers'] ?? []);
+            $options['on_progress'] = self::buildOnProgress(
+                $options['on_progress'] ?? null,
+                [$this->httpParser, 'response'],
+                $spanCustomizer,
+                [$span, 'finish']
+            );
+
+            // Since the cancel event is not being catched by the on_progress
+            // callback we need to manually track it by wrapping it.
+            return new Response(
+                $this->delegate->request($method, $url, $options),
+                $spanCustomizer,
+                [$span, 'finish']
+            );
+        } catch (TransportExceptionInterface $e) {
+            // Since response is an asynchronus operation, according to
+            // HttpClientInterface::request, this exception can only happen
+            // if an unsopported option is passed.
+            $span->tag(ERROR, $e->getMessage());
+            $span->finish();
+            throw $e;
+        }
+    }
+
+    /**
+     * buildOnProgress wraps an existing on_progress to listen to the request
+     * completion. on_progress callback will be called on DNS resolution, on
+     * arrival of headers and on completion; it will also be called on upload/download
+     * of data and at least 1/s
+     */
+    private static function buildOnProgress(
+        ?callable $delegateOnProgress,
+        callable $httpParserResponse,
+        SpanCustomizer $spanCustomizer,
+        callable $spanCloser
+    ): callable {
+        return static function (
+            int $dlNow,
+            int $dlSize,
+            array $info
+        ) use (
+            $delegateOnProgress,
+            $httpParserResponse,
+            $spanCustomizer,
+            $spanCloser
+        ): void {
+            if ($delegateOnProgress !== null) {
+                try {
+                    ($delegateOnProgress)($dlNow, $dlSize, $info);
+                } catch (Throwable $e) {
+                    // According to HttpClientInterface, throwing any exceptions
+                    // MUST abort the request, hence we finish the span in here.
+                    $spanCustomizer->tag(ERROR, $e->getMessage());
+                    ($spanCloser)();
+                    throw $e;
+                }
+            }
+
+            if ($info['error'] !== null || $info['http_code'] > 199) {
+                // any of these three conditions represent the finalization of the request
+                ($httpParserResponse)($dlSize, $info, $spanCustomizer);
+                ($spanCloser)();
+            }
+        };
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function stream($responses, float $timeout = null): ResponseStreamInterface
+    {
+        return $this->delegate->stream($responses, $timeout);
```

Comment:
I'm not sure this works: $responses are going to be instances of `Response`, which delegates doesn't know about

Reply:
So `$this->delegate` is just a wrapped `HttpClientInterface`, my intention here is just to delegate the responsability to that object (hence not do anything special). I might be missing something but I am not sure how that would not work.

Reply:
True. Let me do a couple of changes.

Reply:
Will this do the trick? https://github.com/jcchavezs/zipkin-instrumentation-symfony/pull/65/commits/b4953a305416798a5173cd19aa289fc9750ccf62#diff-46b82fded07ac22b103293cbc8c3ebbeR142